src/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs

```
@@ -86,36 +86,35 @@ public async Task<EmissionsForecast> GetForecastDataAsync(IDictionary props)
         EmissionsForecast forecast;
         using (var activity = Activity.StartActivity())
         {
-            ValidateInput(props);
+            (var location, var forecastRequestedAt) = GetAndValidateForecastInput(props);
+            _logger.LogDebug($"Aggregator getting carbon intensity forecast from data source for location {location} and requestedAt {forecastRequestedAt}");
 
-            var locations = (IEnumerable<Location>)props[CarbonAwareConstants.Locations]!;
-            var forecastRequestedAt = GetOffsetOrDefault(props, CarbonAwareConstants.ForecastRequestedAt, default);
-            _logger.LogDebug($"Aggregator getting carbon intensity forecast from data source for location {locations.First()} and requestedAt {forecastRequestedAt}");
-
-            forecast = await this._dataSource.GetCarbonIntensityForecastAsync(locations.First(), forecastRequestedAt);
+            forecast = await this._dataSource.GetCarbonIntensityForecastAsync(location, forecastRequestedAt);
             var emissionsForecast = ProcessAndValidateForecast(forecast, props);
             return emissionsForecast;
         }
     }
-    private void ValidateInput(IDictionary props)
+    private (Location, DateTimeOffset) GetAndValidateForecastInput(IDictionary props)
```

Reworked this function to reference the `GetXxOrThrow` and use the exception thrown there instead of repeating logic here

---

Wondering if we have made this validation approach more rigid and less extensible. If we were to add additional required parameters, we would have to change the name, signature and implementation of this method.

---

The idea of this was that `GetXxOrThrow` functions already do all the checks we want (null checks, parse checks etc.) so we should leverage them here instead of manually doing it. So one option would be to call for the check, throw away the result and then recall in the main function. Or simply not batch check the forecast inputs and instead have them fail individually (instead of collecting all the errors and then reporting it in one error out)

---

Thanks for the clarification. But does that mean we are not reporting all the errors to the user upfront? I think @bderusha had suggested that it would be a better user experience if they know all the errors in the input payload rather than making multiple requests. Maybe I am missing something.

---

You are correct, I was more giving you the background of this choice versus others we could've made. In the end, we think that this function being less extensible is okay because it's private and is meant to be used by the users of this file. We have another story where we may look at all this validation/defaulting as a whole but for now think it fits the paradigm the best

---

Will resolve for now and we can continue in story 541